src/mdl/isoscelles/isoquant_output_to_scanpy_seurat_input.py

```diff
@@ -0,0 +1,123 @@
+    
+    samfile = pysam.AlignmentFile(bam_path, "rb")
+    read_to_barcode_umi = {}
+    for read in samfile:
+        read_to_barcode_umi.update({read.query_name:(read.get_tag("CB"),read.get_tag("XM"))})
+
+    rname_to_tx = dict()
+    tx_umi_count = defaultdict(lambda: defaultdict(set))
+
+    with open(isoquant_path) as fh:
+        fh.readline()
+        fh.readline()
+        for r in csv.DictReader(fh, delimiter="\t"):
+            if r["#read_id"] in read_to_barcode_umi:
+                if r["assignment_type"] in valid_assignments:
+                    rname_to_tx[r["#read_id"]] = (r["isoform_id"], r["gene_id"])
+                    bc, umi = read_to_barcode_umi[r["#read_id"]]
+                    tx_umi_count[bc][(r["isoform_id"], r["gene_id"])].add(umi)
+
+    if barcode_list is None:
+        barcode_list = sorted(tx_umi_count)
+
+    barcode_index = {bc: i for i, bc in enumerate(barcode_list)}
+
+    if feature_list is None:
+        feature_list = sorted(set(rname_to_tx.values()))
+
+    feature_index = {tx: i for i, tx in enumerate(feature_list)}
+
+    matrix = sparse.COO.from_iter(
+        (
+            ((barcode_index[bc], feature_index[tx]), len(tx_umi_count[bc][tx]))
+            for bc in tx_umi_count
+            for tx in tx_umi_count[bc]
+        ),
+        shape=(len(barcode_list), len(feature_list)),
+        fill_value=0,
+        dtype=int).transpose()
+
+    matrix = pd.DataFrame.sparse.from_spmatrix(matrix)
+    matrix.astype(pd.SparseDtype("float"))
+    
+    if os.path.exists(out_isoform_seurat_path):
+        shutil.rmtree(out_isoform_seurat_path)
+    os.mkdir(out_isoform_seurat_path)
+    
+    with gzip.open(out_isoform_seurat_path + "/matrix.mtx.gz", 'wb') as f:
+        scipy.io.mmwrite(f, scipy.sparse.coo_matrix(matrix))
+    
+    pd.DataFrame(barcode_list).to_csv(out_isoform_seurat_path + "/barcodes.tsv.gz", sep="\t",header=False,index=False, compression='gzip')
+    
+    Ensembl = list(pd.read_csv(gene_Ensembl_to_name, sep='\t', header = None)[0])
+    name = list(pd.read_csv(gene_Ensembl_to_name, sep='\t', header = None)[1])
+    feature_Ensembl = list(pd.DataFrame(feature_list)[1])
+    Enzembl_to_name = {Ensembl[i]: name[i] for i in range(len(Ensembl))}
+    feature_list_pd = pd.DataFrame(feature_list)
+    feature_list_pd[2] = feature_list_pd[1] 
+    feature_list_pd[3] = [Enzembl_to_name[feature_Ensembl[i]] for i in range(len(feature_Ensembl))]
+    feature_list_pd[1] = feature_list_pd[0] + ":" + feature_list_pd[3]
+    feature_list_pd = feature_list_pd[[0,1]]
+    feature_list_pd[2] = "Gene Expression"
+    feature_list_pd.to_csv(out_isoform_seurat_path + "/features.tsv.gz", sep="\t",header=False, index=False, compression='gzip')
+
+    if os.path.exists(out_h5ad_path):
+        shutil.rmtree(out_h5ad_path)
+    os.mkdir(out_h5ad_path)
+
+    anndata = sc.read_10x_mtx(out_isoform_seurat_path)
+    anndata.write_h5ad(out_h5ad_path + "/isoform_feature_bc_matrix.h5ad")
```

Comment:
This is going to be a really slow way to make an `h5ad` file: create a sparse matrix, writes it all out to text, reads it back _in_ and convert back to numbers, and then make the `h5ad` object. It's possible to create an `h5ad` directly from the data.

the `mtx` and `h5ad` formats can be built separately and should live in separate functions, as people might want one or the other and not necessarily both. They can both take the output of `isoquant_matrix` so that the file only needs to be read once.